src/Interpreters/InterpreterAlterQuery.cpp

```diff
@@ -180,6 +181,64 @@ BlockIO InterpreterAlterQuery::executeToTable(const ASTAlterQuery & alter)
+        all_key_columns = merge_tree_table->getAllKeyColumns();
+
+    /// Mutations that should be executed before altering metadata.
+    MutationCommands before_mutation_commands;
+    auto metadata_snapshot = table->getInMemoryMetadataPtr();
+
+    for (auto & command : alter_commands)
+    {
+        /** If we are altering a key column, we should first materialize it.
+          * For instance, when we modify the DEFAULT expression, the order might change (e.g., when the expression depends on another column),
+          * while all other columns expect an order corresponding to the current column default.
+          */
+        if (command.type == AlterCommand::MODIFY_COLUMN && all_key_columns.contains(command.column_name))
+        {
+            if (!command.data_type)
+            {
+                /** If the type is not specified, we don't treat it as a mutation (isRequireMutationStage is set to false, etc.)
+                  * We could update all places to force the mutation stage for such cases, or rewrite the command query ourselves, adding the current type.
+                  * However, this may be error-prone, so we prefer to ask the user to specify the type explicitly, providing a hint in the error message.
+                  */
+                ASTAlterCommand * alter_command = command.ast->as<ASTAlterCommand>();
+                if (alter_command && alter_command->col_decl)
+                {
+                    ASTColumnDeclaration * col_decl = alter_command->col_decl->as<ASTColumnDeclaration>();
+                    for (const auto & current_column : metadata_snapshot->getColumns().getAllPhysical())
+                    {
+                        if (col_decl && current_column.name == command.column_name)
+                        {
+                            col_decl->type = std::make_shared<ASTIdentifier>(current_column.type->getName());
+                            break;
+                        }
+                    }
+                }
+                throw Exception(ErrorCodes::INCORRECT_QUERY, "Cannot MODIFY key column '{}' without specifying new type{}",
+                    command.column_name, alter_command ? fmt::format(", you may try to use '{}' command", alter_command->formatForErrorMessage()) : "");
+            }
+
+            before_mutation_commands.push_back(MutationCommand{.ast = command.ast, .type = MutationCommand::READ_COLUMN, .column_name = command.column_name});
+        }
+    }
+
+    if (before_mutation_commands.hasNonEmptyMutationCommands())
+    {
+        auto context_ = Context::createCopy(getContext());
+        context_->setSetting("mutations_sync", 2);
+        table->checkMutationIsPossible(before_mutation_commands, context_->getSettingsRef());
+        MutationsInterpreter::Settings settings(false);
+        MutationsInterpreter(table, metadata_snapshot, before_mutation_commands, context_, settings).validate();
+        table->mutate(before_mutation_commands, context_);
```

Comment:
Is the sorting also an issue for enum columns?

I'm confused with the divergence of the `tests/queries/0_stateless/01344_alter_enum_partition_key.sql` test: 
```
5,6c5,6
< 3_2_2_1       goodbye 3
< 1_1_1_1       hello   1
---
> 3_3_3_1       goodbye 3
> 1_1_1_1_2     hello   1
10,12c10,12
< 1_1_1_2       1       1
< 111_3_3_1     111     111
< 3_2_2_2       3       3
---
> 1_1_1_2_7     1       1
> 111_8_8_1     111     111
> 3_3_3_2_7     3       3
```
Tried to look at what may be wrong there, don't have an explanation yet.

Reply:
In that test `x` is in partition key, so after the change we will do a mutation to force materialize this column and before `MODIFY COLUMN x`, so mutation id is, so we see different mutation id in this output (and different block numbers because mutation affects it somehow (?))